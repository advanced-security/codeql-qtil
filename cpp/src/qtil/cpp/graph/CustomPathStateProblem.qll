private import qtil.locations.CustomPathStateProblem
private import qtil.cpp.locations.Locatable
private import cpp
// Import the C++ specific configuration for making custom path state problems.
import PathStateProblem<Location, CppLocatableConfig>
