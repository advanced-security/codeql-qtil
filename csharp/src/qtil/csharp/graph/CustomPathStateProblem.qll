private import qtil.locations.CustomPathStateProblem
private import qtil.csharp.locations.Locatable
private import csharp
// Import the C# specific configuration for making custom path state problems.
import PathStateProblem<Location, CsharpLocatableConfig>
