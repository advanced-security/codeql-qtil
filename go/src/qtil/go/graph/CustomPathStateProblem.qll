private import qtil.locations.CustomPathStateProblem
private import qtil.go.locations.Locatable
private import go
// Import the Go specific configuration for making custom path state problems.
import PathStateProblem<go::DbLocation, GoLocatableConfig>
