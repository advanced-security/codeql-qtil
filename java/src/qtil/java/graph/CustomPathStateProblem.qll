private import qtil.locations.CustomPathStateProblem
private import qtil.java.locations.Locatable
private import java
// Import the Java specific configuration for making custom path state problems.
import PathStateProblem<Location, JavaLocatableConfig>
