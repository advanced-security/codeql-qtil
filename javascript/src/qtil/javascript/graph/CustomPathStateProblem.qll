private import qtil.locations.CustomPathStateProblem
private import qtil.javascript.locations.Locatable
private import javascript
// Import the javascript specific configuration for making custom path state problems.
import PathStateProblem<DbLocation, JavascriptLocatableConfig>