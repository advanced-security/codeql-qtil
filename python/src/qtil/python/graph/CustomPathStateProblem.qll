private import qtil.locations.CustomPathStateProblem
private import qtil.python.locations.Locatable
private import python
// Import the python specific configuration for making custom path state problems.
import PathStateProblem<Location, PythonLocatableConfig>
