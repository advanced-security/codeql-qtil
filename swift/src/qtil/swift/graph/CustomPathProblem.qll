private import qtil.locations.CustomPathProblem
private import qtil.swift.locations.Locatable
private import swift
// Import the swift specific configuration for making custom path problems.
import PathProblem<Location, SwiftLocatableConfig>
