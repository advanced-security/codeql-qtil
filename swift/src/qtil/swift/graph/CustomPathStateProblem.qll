private import qtil.locations.CustomPathStateProblem
private import qtil.swift.locations.Locatable
private import swift
// Import the swift specific configuration for making custom path state problems.
import PathStateProblem<Location, SwiftLocatableConfig>
