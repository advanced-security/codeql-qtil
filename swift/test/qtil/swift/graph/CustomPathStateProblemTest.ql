// Minimal test to make sure CustomPathStateProblem works with Swift code.
import swift
import qtil.swift.graph.CustomPathStateProblem

module CallGraphPathStateProblemConfig implements CustomPathStateProblemConfigSig {
  class Node = VarDecl;

  class State = int; // Track search depth

  predicate start(Node n, int depth) { n.getName() = "start" and depth = 0 }

  bindingset[depth]
  predicate end(Node n, int depth) { n.getName() = "end" }

  bindingset[depth1]
  bindingset[depth2]
  predicate edge(VarDecl a, int depth1, VarDecl b, int depth2) {
    depth2 = depth1 + 1 and
    exists(PatternBindingDecl bindingA, PatternBindingDecl bindingB |
      bindingA.getPattern(0).(NamedPattern).getVarDecl() = a and
      bindingB.getPattern(0).(NamedPattern).getVarDecl() = b and
      bindingA.getInit(0).(DeclRefExpr).getDecl() = b
    )
  }
}

import CustomPathStateProblem<CallGraphPathStateProblemConfig>

from VarDecl start, VarDecl end
where problem(start, end)
select start, end, "Path from $@ to $@.", start.getName(), start, end.getName(), end
