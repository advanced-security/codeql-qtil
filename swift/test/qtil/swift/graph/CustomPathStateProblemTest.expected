WARNING: unused variable 'depth' (CustomPathStateProblemTest.ql:13,29-34)
edges
nodes
#select